shared/lib/neighbors/nAryHeap.ts

```diff
@@ -0,0 +1,108 @@
+/**
+*  @license
+* Copyright 2021, JsData. All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ==========================================================================
+*/
+
+const ARITY = 8
+const child = (parent: number) => parent * ARITY + 1
+const parent = (child: number) => ((child - 1) / ARITY) | 0
+
```

Comment:
Why are we rolling our own nAryHeap? Should we just use an off the shelf library?

Reply:
* I am hesitant to add a dependency on something that is fairly easily implementable (trying to not end up with the [is-even](https://www.npmjs.com/package/is-even) package, I guess :laughing:)
  * at a glance, available binary-heap/priority-queue packages did not inspire enough confidence
  * somewhere along the way I was considering that, when using Tensors in a priority queue, said queue should be async. A custom implementation could be changed to support that.
  * n-ary heaps can trade-off `popMin()` performance for improved `add(x)` performance, which makes sense in the case of kd-tree, where there are many more adds than pops.